007-extension-flag.md

```diff
@@ -0,0 +1,57 @@
+| Title  | Implement --extension flag  |
+|--------|-----------------------------|
+| Author | @bmeck                      |
+| Status | DRAFT                       |
+| Date   | 2017-08-09T10:00:00-05:00   |
+
+## Description
+
+The Node CLI currently accepts input in both filename (`node app.js`) and opaque text stream form (`node -e '123'` and `node <app.js`). This behavior always results in running the stream as a CommonJS goal. With the upcoming usage of ESM and WASM there will be cases where developers wish to pipe other forms of text to the CLI.
+
+This proposal is to implement an `--entry-extension` CLI flag for the cases which may want to use a stream that does not have a file extension.
+
+## Example
+
+```sh
+node --entry-extension=js <app.js
+node --entry-extension=js -e $(cat <app.js)
+node --entry-extension=wasm <app.wasm
+node --entry-extension=wasm -e $(cat <app.wasm)
+node --entry-extension=mjs <app.mjs
+node --entry-extension=mjs -e $(cat <app.mjs)
+```
+
+## Considerations
+
+If mixing `--entry-extension` with a non-stream argument such as a file path or entry URL, Node should exit with an error prior to any evaluation. This also means that `--entry-extension` cannot be used to override the extension of file paths.
+
+```sh
+# exits with an error
+node --entry-extension=json app.js
+# exits with an error
+node --entry-extension=js app.js
+```
+
+### Userland Extensions
+
+Some libraries extend `require.extensions` with new extensions. These libraries should be considered when determining the list of extensions allowed in `--entry-extension`.
+
+Any future extensions to the ESM based loading mechanisms will have to account for collisions and define behavior when they collide.
+
+### Early Errors
+
+Any unknown file extension should prevent the main entry point to the program from executing anything. Preloaded files such as using `--require` should continue to load normally.
```

Comment:
Is there any concept of preloading files, but with a different entry extension? In other words, `--import` to match `--require`?

Reply:
@ljharb not currently, use a wrapper for now. `--require` uses file paths always, so file extensions are always included. There are some odd problems with naming here, and some timing differences due to ESM being async that I think are fine to put off for a future time.

Reply:
ah, so `--require file.mjs` would work as expected?

Reply:
maybe? `require('./foo.mjs')` throws so the naming is a bit odd but since CLI arg for main file to run accepts `.mjs` it should work fine I would expect.

Reply:
Correct, but `require()` has synchronous code boundaries to deal with. We have more freedom in how we define preload's interactions.

Perhaps we should have `--import` alias to `--require` though, for posterity and to avoid confusion. That being said, `-i` is already used.

Reply:
IDK, `--import` seems fine and can use the module resolution of ESM rather than CJS *then* ESM